src/client/common/alias_dict.cpp

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (C) 2021 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <multipass/cli/alias_dict.h>
+#include <multipass/constants.h>
+#include <multipass/file_ops.h>
+#include <multipass/format.h>
+#include <multipass/json_writer.h>
+#include <multipass/standard_paths.h>
+#include <multipass/utils.h>
+
+#include <QDir>
+#include <QFile>
+#include <QJsonArray>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QJsonParseError>
+#include <QTemporaryFile>
+
+namespace mp = multipass;
+namespace mpu = multipass::utils;
+
+mp::AliasDict::AliasDict()
+{
+    const auto file_name = QStringLiteral("%1_aliases.json").arg(mp::client_name);
+    const auto user_config_path = QDir{MP_STDPATHS.writableLocation(mp::StandardPaths::GenericConfigLocation)};
+    const auto cli_client_dir_path = QDir{user_config_path.absoluteFilePath(mp::client_name)};
+
+    aliases_file = cli_client_dir_path.absoluteFilePath(file_name).toStdString();
+
+    load_dict();
+}
+
+mp::AliasDict::~AliasDict()
+{
+    if (modified)
+    {
+        save_dict();
+    }
+}
+
+void mp::AliasDict::add_alias(const std::string& alias, const mp::AliasDefinition& command)
+{
+    if (aliases.try_emplace(alias, command).second)
+    {
+        modified = true;
+    }
+}
+
+bool mp::AliasDict::remove_alias(const std::string& alias)
+{
+    if (aliases.erase(alias) > 0)
+    {
+        modified = true;
+        return true;
+    }
+
+    return false;
+}
+
+void mp::AliasDict::remove_aliases_for_instance(const std::string& instance)
+{
+    for (auto it = aliases.begin(); it != aliases.end();)
+    {
+        if (it->second.instance == instance)
+        {
+            modified = true;
+            it = aliases.erase(it);
+        }
+        else
+            ++it;
+    }
+}
+
+mp::optional<mp::AliasDefinition> mp::AliasDict::get_alias(const std::string& alias) const
+{
+    try
+    {
+        return aliases.at(alias);
+    }
+    catch (const std::out_of_range&)
+    {
+        return mp::nullopt;
+    }
+}
+
+void mp::AliasDict::load_dict()
+{
+    QFile db_file{QString::fromStdString(aliases_file)};
+
+    aliases.clear();
+
+    if (db_file.exists())
+    {
+        if (!db_file.open(QIODevice::ReadOnly))
+            throw std::runtime_error(fmt::format("Error opening file '{}'", db_file.fileName()));
```

Comment:
You should be able to make a unit test for this if you change this to use the `FileOps` singleton class.

Reply:
Done.